File: elixir/meetup/meetup.exs
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @days [ :monday, :tuesday, :wednesday,
          :thursday, :friday, :saturday, :sunday ]

  @type weekday ::
      :monday | :tuesday | :wednesday
    | :thursday | :friday | :saturday | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: :calendar.date
  def meetup(year, month, weekday, schedule) do
    1..31
    |> Enum.map(&({ year, month, &1 }))
    |> Enum.filter(&(:calendar.valid_date(&1)))
    |> Enum.filter(&(same_day_number?(&1, weekday)))
    |> Enum.filter(&(same_schedule?(&1, schedule)))
    |> List.last
  end

  @spec same_day_number?(:calendar.date, weekday) :: boolean
  defp same_day_number?(date, day) do
    :calendar.day_of_the_week(date) === Enum.find_index(@days, &(&1 === day)) + 1
  end

  @spec same_schedule?(:calendar.date, schedule) :: boolean
  defp same_schedule?({ _, _, day }, :first), do: day <= 7
  defp same_schedule?({ _, _, day }, :second), do: day in 8..14
  defp same_schedule?({ _, _, day }, :third), do: day in 15..21
  defp same_schedule?({ _, _, day }, :fourth), do: day in 22..28
  defp same_schedule?({ _, _, day }, :last), do: true
  defp same_schedule?({ _, _, day }, :teenth), do: day in 13..19
  defp same_schedule?(_, _), do: false 
end

File: ecmascript/meetup/meetup.js
const DAYS = [
  'Sunday', 'Monday', 'Tuesday', 'Wednesday',
  'Thursday', 'Friday', 'Saturday'
];

const meetupDay = (y, m, d, s) => {
  const numDays = new Date(y, m + 1, 0).getDate();
  const candidates = Array(numDays).fill()
    .map((_, i) => i + 1)
    .filter(day => new Date(y, m, day).getDay() == DAYS.indexOf(d));
  const date = matchSchedule(candidates, s);

  if (!date) throw new Error;

  return new Date(y, m, date);
};

const matchSchedule = (candidates, schedule) => {
  return {
    '1st': candidates[0],
    '2nd': candidates[1],
    '3rd': candidates[2],
    '4th': candidates[3],
    '5th': candidates[4],
    'last': candidates[candidates.length - 1],
    'teenth': candidates.find(d => d >= 13 && d <= 19)
  }[schedule];
};

export default meetupDay;
